lab_2/main.py

```diff
@@ -123,4 +280,27 @@ def predict_language_knn_sparse(unknown_text_vector: list, known_text_vectors: l
     :param language_labels: language labels for each known text
     :param k: the number of neighbors to choose label from
     """
-    pass
+    if not (
+        isinstance(unknown_text_vector, list)
+        and all(isinstance(n, list) for n in unknown_text_vector)
+        and isinstance(known_text_vectors, list)
+        and all(isinstance(m, list) for m in known_text_vectors)
+        and isinstance(language_labels, list)
+        and all(isinstance(s, str) for s in language_labels)
+        and len(known_text_vectors) == len(language_labels)
+        ):
+        return None
+    distances = []
+    for known_text_vector in known_text_vectors:
+        distance = calculate_distance_sparse(unknown_text_vector, known_text_vector)
+        distances.append(distance)
+    sorted_distances = (sorted(distances))[:k]
+    langs = []
+    for i in sorted_distances:
+        index = distances.index(i)
+        langs.append(language_labels[index])
+    langs_counter = {}
+    for lang in set(langs):
+        langs_counter[lang] = langs.count(lang)
+    knn_predict = [max(langs_counter, key=langs_counter.get), min(sorted_distances)]
+    return knn_predict
```

Comment:
can you describe in writing what happens in this function?

Reply:
303. 'for' loop starts.
set() is used to remove duplicate items from the list of languages (langs).
e. g. set(['de', 'de', 'en']) -> {'en', 'de'}

304. Add elements in the dictionary langs_counter, where key is a language label from set(langs), value is the number of repetitions of this label in the list of languages.
e. g. langs_counter = {'en': 1, 'de': 2}

305. Create a list with a language label and a value - the minimum distance to the nearest neighbour.
The list consists of 2 elements - language label and minimum distance.
We find an appropriate language label in the dictionary langs_counter by getting the key with the maximum value. 
We find a minimum distance by finding the minimum value in a list of k closest distances (sorted_distances).

306. Return the list